server/channels/jobs/batch_migration_worker.go

```diff
@@ -0,0 +1,234 @@
+	worker.logger.Debug("Worker stopping")
+	close(worker.stop)
+	<-worker.stopped
+}
+
+// JobChannel is the means by which the jobs infrastructure provides the worker the job to execute.
+func (worker *BatchMigrationWorker) JobChannel() chan<- model.Job {
+	return worker.jobs
+}
+
+// IsEnabled is always true for batch migrations.
+func (worker *BatchMigrationWorker) IsEnabled(_ *model.Config) bool {
+	return true
+}
+
+// checkIsClusterInSync returns true if all nodes in the cluster are running the same version,
+// logging a warning on the first mismatch found.
+func (worker *BatchMigrationWorker) checkIsClusterInSync() bool {
+	clusterStatus := worker.app.GetClusterStatus()
+	for i := 1; i < len(clusterStatus); i++ {
+		if clusterStatus[i].SchemaVersion != clusterStatus[0].SchemaVersion {
+			worker.logger.Warn("Worker: cluster not in sync", mlog.String("schema_version_a", clusterStatus[0].SchemaVersion), mlog.String("schema_version_b", clusterStatus[1].SchemaVersion))
+			return false
+		}
+	}
+
+	return true
+}
+
+// DoJob executes the job picked up through the job channel.
+//
+// Note that this is a lot of distracting machinery here to claim the job, then double check the
+// status, and keep the status up to date in line with job infrastrcuture semantics. Unless an
+// error occurs, this worker should hold onto the job until its completed.
+func (worker *BatchMigrationWorker) DoJob(job *model.Job) {
+	logger := worker.logger.With(mlog.Any("job", job))
+	logger.Debug("Worker received a new candidate job.")
+	defer worker.jobServer.HandleJobPanic(logger, job)
+
+	if claimed, err := worker.jobServer.ClaimJob(job); err != nil {
+		logger.Warn("Worker experienced an error while trying to claim job", mlog.Err(err))
+		return
+	} else if !claimed {
+		return
+	}
+
+	c := request.EmptyContext(logger)
+	var appErr *model.AppError
+
+	// We get the job again because ClaimJob changes the job status.
+	job, appErr = worker.jobServer.GetJob(c, job.Id)
+	if appErr != nil {
+		worker.logger.Error("Worker: job execution error", mlog.Err(appErr))
+		worker.setJobError(logger, job, appErr)
+		return
+	}
+
+	if job.Data == nil {
+		job.Data = make(model.StringMap)
+	}
+
+	for {
+		select {
+		case <-worker.stop:
+			logger.Info("Worker: Migration has been canceled via Worker Stop. Setting the job back to pending.")
+			if err := worker.jobServer.SetJobPending(job); err != nil {
+				worker.logger.Error("Worker: Failed to mark job as pending", mlog.Err(err))
+			}
+			return
+		case <-time.After(worker.timeBetweenBatches):
+			// Ensure the cluster remains in sync, otherwise we restart the job to
+			// ensure a complete migration. Technically, the cluster could go out of
+			// sync briefly within a batch, but we accept that risk.
+			if !worker.checkIsClusterInSync() {
+				worker.logger.Warn("Worker: Resetting job")
+				worker.resetJob(logger, job)
+				return
+			}
+
+			worker.logger.Trace("Worker: processing batch", mlog.Any("data", job.Data))
```

Comment:
Would suggest to keep this at debug to keep things simple. The expectation is that except the trace settings we have for things like Elasticsearch, LDAP, setting log level to debug will log everything.

Reply:
Ack, re: keeping things simple. I'm inclined to remove this log altogether given the potential for log spamming, especially for migrations that one day touch millions of rows one day.

Reply:
This will be called once per batch right? 0/5 but this might still be useful in cases where the job just gets randomly stuck. But not a blocker from my side.